modules/imgcodecs/src/grfmt_jpeg.cpp

```diff
@@ -253,12 +255,73 @@ bool  JpegDecoder::readHeader()
         }
     }
 
+    m_orientation = getOrientation();
+
     if( !result )
         close();
 
     return result;
 }
 
+int JpegDecoder::getOrientation()
+{
+    int orientation = JPEG_ORIENTATION_TL;
+
+    ExifReader reader( m_filename );
+    if( reader.parse() )
+    {
+        orientation = reader.getTag( ORIENTATION ).field_u16;//orientation is unsigned short, so check field_u16
+    }
+
+    return orientation;
+}
+
+void JpegDecoder::setOrientation(Mat& img)
+{
+    Mat dst = img; //I didn't investigate deeply but making just declaration Mat dist;
+                    //(and calling the Mat's default constructor) instead of initialization
+                    //by copy img leads to failing the test
+                    //opencv_perf_stitching --gtest_filter=OCL_stitch_boat.boat/0
+                    //CV_Assert( ssize.area() > 0 ); in modules/imgproc/src/imgwarp.cpp, line 3233
+                    //So there are two outcomes of initializing dst with img:
+                    //1. Getting rid of test failing
+                    //2. Handling properly the default case in switch
```

Comment:
Well, you can remove `dst` totally from the current version of code and use `img` only.

Test had been failed before because it "reads" empty Mat as result (`m_orientation` was 0 and it was not equal `JPEG_ORIENTATION_TL` which is 1).